### PR TITLE
Create thumbnails for episodes

### PR DESCRIPTION
**Pull request by theborakompanioni**
Currently episode images are displayed "as is" -  images with size 3000x3000 and >1MB for showing them at 128x128 (past episodes) and 160x160 (most recent episode or player image) resolution.

This commit fetches the images and saves a thumbnail into the public asset directory.
An episode thumbnail is 160x160 and ~10KB in size with the current settings.

Pro:
- page load increase - happy users

Cons:
- Build time increase (synchronous filesystems checks) - sad devs
- Repo size increase (negligible)

I guess there are quite a few improvements to be made, hence it is a draft PR for now. 
Feedback welcome.

### CODE REVIEW

**Review by dergigi — COMMENTED** [review]
Looks good at a first glance.

I think if we're going for this we should go "all the way" so to speak and don't generate `jpg` files but `webp` files. Imagemagick supports this natively, at least the cli version. We can also generate both to remain backwards compatible with old browsers, and have the jpgs as a fallback like [this](https://usefulangle.com/post/114/webp-image-in-html-with-fallback).

**Comment by theborakompanioni** [comment]
> I think if we're going for this we should go "all the way" so to speak and don't generate `jpg` files but `webp` files.

Done!

Size of 160x160 is okay?

**Review by dnlggr — DISMISSED** [review]
Nice, good idea! That was on my bucket list for quite a while now! 😄 

Some feedback on the general approach:

Next.js has [image optimization features](https://nextjs.org/docs/basic-features/image-optimization) built in already. The only issue is that we are using the [static export feature](https://nextjs.org/docs/advanced-features/static-html-export) which means that image optimization doesn't work out of the box. We don't really need to run the site as a static webpage but for now it's probably also not worth to change that (although out of the box image optimization could be a reason to do so maybe.)

The docs mention that image optimization should work when using a [_custom loader_](https://nextjs.org/docs/advanced-features/static-html-export#supported-features) (whatever that is). I never got around to dig deeper but it sounds like there could be a way of using Nextjs's optimization stuff even for a static site like ours. Do you know more about that? If it turns out to be a big hassle, let's just keep the custom stuff but I feel like it could be worth trying.

Basically what I'm saying is let's try if we can get the build in optimization stuff to work with our static site but it it is a lot of work, let's keep what you already have—does that make sense?

---

As for the implementation itself: Nice! I think the extra couple of seconds build time are definitely worth it. One thing I noticed is that the images are compressed a bit too much. When you view the site on a high res display they are really really blurry. 😉  Let's maybe increase the resolution a bit. And a second thing: The image dimensions  are a bit off I believe. I noticed it when looking at the cards because the padding wasn't equal on all sides anymore:

**Before:**
Screenshot 2022-01-27 at 13 56 18

**After:**
Screenshot 2022-01-27 at 13 56 29

Those last two things (blurry images and dimensions) are my change requests, the rest is just thinking out loud. Let me know what you think!

**Comment by theborakompanioni** [comment]
Nice, thank you @dnlggr. This is such amazing feedback!
Good catch with the dimensions - I certainly would not have noticed..

> Next.js has [image optimization features](https://nextjs.org/docs/basic-features/image-optimization) built in already. [...] Do you know more about that?

No, I know nothing about image optimization in nextjs yet. I will definitely read into it and will adapt the code and report my findings.

¡Muchas gracias!

**Comment by dnlggr** [comment]
> will adapt the code 

It's very well possible that it doesn't work or only works with a lot of hassle with statically exported sites. So if it is too much of a time sink, the manual approach is fine too!

**Comment by theborakompanioni** [comment]
> > will adapt the code
> 
> It's very well possible that it doesn't work or only works with a lot of hassle with statically exported sites. So as I said if it is too much of a time sink, the manual approach is fine too!

You were right again, it really is a lot of hassle. After many attempts, I could not get it to work properly.

See discussion here: https://github.com/vercel/next.js/discussions/19065
But I could not get the proposed solution with `next-optimized-images` to work, as many others.
Also:
> I would warn against anyone using next-optimized-images since it's not being maintained. You'll be stuck on next.js v11 for the foreseeable future. - https://github.com/vercel/next.js/discussions/19065#discussioncomment-2013457

See also their [proposed improved "wording" to in their documentation](https://github.com/vercel/next.js/pull/33555/files).

But all in all I think the current approach of downloading and rescaling the images can be improved with some of the new things I came across thanks to your links. Hope to have it by end of the week.

**Comment by dnlggr** [comment]
> You were right again, it really is a lot of hassle. After many attempts, I could not get it to work properly.

Damit, I was hoping a "custom loader" could be a callback that downloads the original image and next does the rest...

Well, then let's stick with the custom approach. We just need to make sure the images are still crisp on higher resolution displays. 🖼️ ✨

**Comment by theborakompanioni** [comment]
> > will adapt the code
> 
> It's very well possible that it doesn't work or only works with a lot of hassle with statically exported sites. So as I said if it is too much of a time sink, the manual approach is fine too!

You were right again, it really is a lot of hassle. After many attempts, I could not get it to work properly.

See discussion here: https://github.com/vercel/next.js/discussions/19065
But I could not get the proposed solution with `next-optimized-images` to work, as many others.
Also:
> I would warn against anyone using next-optimized-images since it's not being maintained. You'll be stuck on next.js v11 for the foreseeable future. - https://github.com/vercel/next.js/discussions/19065#discussioncomment-2013457

See also their [proposed improved "wording" to in their documentation](https://github.com/vercel/next.js/pull/33555/files).

But all in all I think the current approach of downloading and rescaling the images can be improved with some of the new things I came across thanks to your links. Will push an update soon.

**Comment by theborakompanioni** [comment]
Changes:
- thumbnail size to 400x400
- use squoosh instead of imagemagick
- download/transforming images in bulk (while preventing massive parallel downloads)
- hopefully all visual discrepancies fixed

Will squash into single commit after review.

File: src/lib/thumbnails.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+export async function downloadAllMissingEpisodeThumbnails(episodes, size = default_size) {
+  const imagePool = new ImagePool()
+
+  for (const episode of episodes) {
+    const thumbnailFileNamesByFormat = toThumbnailFileNamesByFormat(episode, size, supported_thumbnail_formats)
+
+    const allThumbnailsPresent = await Object.values(thumbnailFileNamesByFormat)
+      .map(async (fileName) => await fileExists(`${episode_thumbnail_dir}/${fileName}`))
+      .reduce((a, b) => a && b, true)
```
Comment (dnlggr) [inline src/lib/thumbnails.js:54]: The `map` call returns an array of promises here so you need to wait for them inside reduce. Or wait for them all at once and then reduce on the resolved results.

I think something like this should work. It's maybe not the solution that's easiest on the eye. Just to illustrate what I mean:

```suggestion
      .reduce(async (a, b) => (await a) && (await b), true)
```

**Comment by dnlggr** [comment]
Cool, almost there! Images look good now. 👌 

See my inline comments, Nothing big, let's merge this today! 😄

File: src/lib/thumbnails.js (hunk 1 above)
Comment (theborakompanioni) [reply src/lib/thumbnails.js:54]: You are completely right. Strange, that it does what it is supposed to..

Would
```js
const allThumbnailsPresent = await Promise.all(Object.values(thumbnailFileNamesByFormat)
      .map(async (fileName) => await fileExists(`${episode_thumbnail_dir}/${fileName}`)))
      .reduce((a, b) => a && b, true)
```
be an acceptable adaption?

File: src/lib/feed.js [hunk 2]
```diff
@@ -136,5 +136,12 @@ export async function fetchEpisodes() {
     return episode
   })
 
+  await downloadAllMissingEpisodeThumbnails(episodes)
+
+  episodes.forEach((episode) => {
+    episode.thumbnails = thumbnailUrlsByFormat(episode)
+    episode.thumbnailFallback = episode.thumbnails.jpg
```
[the remark this answers is not in the document]
Comment (theborakompanioni) [reply src/lib/feed.js:143]: Yes, that was on purpose as it should not be possible to build successfully without the images being present. But you have a point, we might as well use the original! Will change!

File: src/lib/thumbnails.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
+import fs from 'fs'
+import { ImagePool } from '@squoosh/lib'
+
+const episode_thumbnail_dir = 'public/assets/podcast/thumbnails'
+const episode_thumbnails_url_path = '/assets/podcast/thumbnails'
+
+const supported_thumbnail_formats = ['jpg', 'webp']
+const default_size = 400
```
[the remark this answers is not in the document]
Comment (theborakompanioni) [reply src/lib/thumbnails.js:8]: Having sneak_case was the result of trying to be consistent. I saw `boostagrams_dir` in [lib/boostagrams.js#L6](https://github.com/seetee-io/seetee-io/blob/master/src/lib/boostagrams.js#L6). This is not meant to be a justification, just an explanation. 

Will change to camelCase!

File: src/lib/thumbnails.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+const toThumbnailFileNamesByFormat = (episode, size, formats) =>
+  Object.fromEntries(formats.map((format) => [format, toEpisodeThumbnailFileName(episode, size, format)]))
+
+const fileExists = async (file) =>
+  fs.promises
+    .access(file)
+    .then(() => true)
+    .catch(() => false)
+
```
Comment (dnlggr) [inline src/lib/thumbnails.js:21]:
```suggestion
```

**Review by dnlggr — APPROVED** [review]
LGTM! Good improvement! 👏👏

**Review by dnlggr — APPROVED** [review]

**Comment by dnlggr** [comment]
Loads mich faster now—nice! 👏

**Pull request merged**